images/cilium/Dockerfile

```diff
@@ -93,3 +93,20 @@ WORKDIR /home/cilium
 
 ENV INITSYSTEM="SYSTEMD"
 CMD ["/usr/bin/cilium"]
+
+#
+# Cilium debug image.
+#
+# Typical image bulids will stop above at the 'release' target, but
+# developers follow this Dockerfile to the end. Starting from a release
+# image, install delve debugger and wrap the cilium-agent binary calls
+# with a script that automatically provisions the debugger on a
+# dedicated port.
+FROM release as debug
+# TARGETOS is an automatic platform ARG enabled by Docker BuildKit.
+ARG TARGETOS
+# TARGETARCH is an automatic platform ARG enabled by Docker BuildKit.
+ARG TARGETARCH
+COPY --from=builder /go/bin/dlv /usr/bin/dlv
+RUN mv /usr/bin/cilium-agent /usr/bin/cilium-agent-bin
+COPY images/scripts/debug-wrapper.sh /usr/bin/cilium-agent
```

Comment:
You technically don't need to do this whole switch array. 

You can also just make the wrapper script look like this:

```
set -o errexit
set -o nounset
set -o pipefail

/usr/bin/dlv \
    --listen=:2345 \
    --log=true \
    --log-output=debugger,debuglineerr,gdbwire,lldbout,rpc \
    --accept-multiclient \
    --api-version=2 \
    dap
```

And the launch.json look like:

```
    "configurations": [
		{
			"name": "Attach to kind-control-plane-1",
			"type": "go",
			"request": "launch",
			"mode": "exec",
			"program": "/usr/bin/cilium-agent",
			"port": 23401,
			"host":"127.0.0.1",
			"showLog": true,
			"trace": "log",
			"logOutput": "rpc",
			"debugAdapter": "dlv-dap",
			"substitutePath": [{"from": "${workspaceFolder}", "to": "/go/src/github.com/cilium/cilium"}],
		},
```

The difference here is that delve is started as a dap server, but doesn't do anything until the client connects. When it does, the client tells Delve to launch cilium-agent. 

This **would** be beneficial, since you could ask delve to "rerun" and it could restart the binary, but unfortunately I think that is currently broken in Delve anyway. 

So i'll let you judge which route is better.

Reply:
> The difference here is that delve is started as a dap server, but doesn't do anything until the client connects. When it does, the client tells Delve to launch cilium-agent.

:thinking: that's exactly the behaviour I was observing. If I follow right, what I do is move the real binary out of the way, move cilium-agent into the place where the regular binary would be, then when kubelet runs the pod, it just runs the $thing at /usr/bin/cilium-agent which _happens_ to be the wrapper script. Since it's the wrapper script, that will start the debugger servers (in this case, both DAP and Delve API v2), then it will start running the agent with standard args but pause it before any logic runs. With the current approach, I'm not modifying the Pod / DaemonSet YAMLs at all, they're stock standard. Incidentally, the wrapper script should also work for cilium-operator in the same way since there's no direct references to the binary.

I *think* you're suggesting to modify the pod YAML so that the entrypoint to the container is the path to this debugger script, which would also work. How do we customize the args to hand to dlv in this case? One thing I want to make sure happens is that we pass the same config flags as what was already in the pod, so that the configuration is properly implemented (ie it's not like just by running the debugger you randomly change settings that could impact reproducibility of some bug).

> This would be beneficial, since you could ask delve to "rerun" and it could restart the binary, but unfortunately I think that is currently broken in Delve anyway. 

I was doing this with this approach (clicking the button to restart the app), it seemed to work already :sweat: But maybe it doesn't work 100% the same as what you were saying - I think I noticed that the error counts would increase in the `kubectl get pods` output. So maybe I was just observing that the binary would exit with some error code when I click the restart button, then maybe I was relying on kubelet to restart the application (which it will happily do, though after a couple of times it may introduce CrashLoopBackoff timeouts).

Reply:
^^ Yeah confirmed this, I've been relying on k8s to restart the app, so if you do this too many times it starts to introduce arbitrary delays because it thinks that the app is failing too much. It's not a huge deal unless you're needing to constantly restart the agent to debug init logic. Not sure what's the best path to mitigate that at this point. Part of me wants to adopt your suggestion, I'm just not sure about a clean way to introduce this into the DaemonSet to keep the changes minimal vs. the real production DaemonSet.

Reply:
You know, now that you mention it, it would be nicer that we don't need to modify the daemonset at all. 

Just to be clear, I wasn't "suggesting" the above, just showing another way of going about it, in which the cilium-agent/cilium-agent-bin swap around wasn't strictly necessary. 

> ^^ Yeah confirmed this, I've been relying on k8s to restart the app, so if you do this too many times it starts to introduce arbitrary delays because it thinks that the app is failing too much. It's not a huge deal unless you're needing to constantly restart the agent to debug init logic. Not sure what's the best path to mitigate that at this point. Part of me wants to adopt your suggestion, I'm just not sure about a clean way to introduce this into the DaemonSet to keep the changes minimal vs. the real production DaemonSet.

Here's the kicker, I don't believe there is a way around this since Delve does not support the "Restart" DAP command. So we can't ask DAP to restart the application its been configured to launch. Instead, when it sees a disconnect it will just terminate the launched program and exit itself. 

It also appears there are no plans to support this: https://github.com/go-delve/delve/discussions/3049 (with a little comment from your's truly lolol).

Anyway, the above comments were just food for thought, but now that you explained why you went about it this way, and the Daemon set entry point considerations, I think the way we have it in this PR is probably the best.